os/fs/smartfs/smartfs_smart.c

```diff
@@ -289,26 +289,11 @@ static int smartfs_open(FAR struct file *filep, const char *relpath, int oflags,
 			}
 
 			/* First try to find an invalid or empty entry available in one of the chained parent sectors */
-			ret = smartfs_find_availableentry(fs, parentdirsector, &newentry);
+			ret = smartfs_find_availableentry(fs, parentdirsector, &newentry, &new_chain);
```

Comment:
Ohkay, similar to your comment on #4406 
Let's check how this can be done without using an extra variable

Reply:
Earlier,
```
ret = smartfs_find_availableentry();
if (ret == -ENOENT) {
        new_chain = TRUE;
        smartfs_createentry();
}
```
If no available entry is found, we automatically know a new sector will be chained.
createentry is called from smartfs_smart.c i.e. top of fs layer itself

But now,
```
int smartfs_availableentry(..................)
{
        //try to find an entry
        //if found, good
        //not found?
                smartfs_createentry();
}
```
The call to smartfs_createentry is now implicit from smartfs_availableentry().
So top of fs layer (smartfs_smart.c) cannot know if new chain was added or not because it is not calling createentry directly.

Yes, as you said parentdirsecor is passed to find_availableentry to find free entry.
But consider this

parentdirsector = 24
**24** -> **25** -> **26** -> 0xFF
No availableentry found, so new sector is chained after 26, say sector 27
Function returns with newentry->dsector = 27.

```
If (parentdirsector != newentry->dsector)
        new_chain = TRUE;
```

But, what if an entry is found in sector 25 or 26,
In this case also, parentdirsector != newentry->dsector
so new_chain = TRUE which is wrong

So we will have to check entire sector chain of parentdirsector
```
while (parentdirsector != 0xFF) {
        if (parentdirsector == newentry->dsector) {
                break;

        //Read the current sector into buffer
        parentdirsector = parentdirsector -> nextsector
        // Continue the loop
}
```

I think this loop check  is a lot worse than passing one new variable to the function...

Reply:
In the above case, when newentry->dsector = 27.
Still we don't know if 27 was newly allocated or present already in the chain.

We will have to go till the end of the chain of parentdirsector,
Then check if that sector == newentry->dsector
AND also if newentry->doffset = sizeof(chain header)

Then MAYBE, it is a  new sector
It may be the case that the first entry in 27 was found available because it was invalidated earlier.
In this case, we cannot determine whether 27 is a new sector or old one.

Reply:
In my opinion, because it is tightly related to this PR & 4406, we'd better send parentdirsector.
because I don't agree asign sector number to offset....